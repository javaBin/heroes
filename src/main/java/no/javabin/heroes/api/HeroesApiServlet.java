package no.javabin.heroes.api;

import java.io.IOException;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import no.javabin.heroes.Profile;
import no.javabin.heroes.hero.HeroesContext;
import no.javabin.infrastructure.http.server.ApiServlet;
import org.fluentjdbc.DbContext;
import org.fluentjdbc.DbContextConnection;

public class HeroesApiServlet extends ApiServlet {

    private final HeroesContext heroesContext;
    private DbContext dbContext;

    public HeroesApiServlet(HeroesContext heroesContext) {
        this.heroesContext = heroesContext;
        this.dbContext = new DbContext();
    }

    @Override
    protected void service(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
        try(DbContextConnection connection = dbContext.startConnection(heroesContext.getDataSource())) {
            super.service(req, resp);
        }
    }

    @Override
    public void init() throws ServletException {
        registerController(new LoginController(heroesContext, dbContext));
        registerController(new AdminController(dbContext));
        registerController(new ProfileController(dbContext));
    }

    @Override
    protected boolean isUserLoggedIn(HttpServletRequest req) {
        return getProfile(req) != null;
    }

    @Override
    protected boolean isUserInRole(HttpServletRequest req, String role) {
        if (role.equals("admin")) {
            return getProfile(req).hasAdminScope();
        }
        return false;
    }

    private Profile getProfile(HttpServletRequest req) {
        return (Profile) req.getSession().getAttribute("profile");
    }
}
