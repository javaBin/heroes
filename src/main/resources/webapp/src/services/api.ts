
export interface Person {
    name: string;
    email: string;
    twitter?: string;
}

export interface HeroAchievement {
    id?: string;
    type: Achievement;
    label?: string;
}

export interface Hero extends Person {
    avatar?: string;
    achievements: HeroAchievementDetail[];
    id?: string;
    achievement?: string;
    published?: boolean;
}

export interface Userinfo {
    authenticated?: boolean;
    admin?: boolean;
    username?: string;
}

export type Achievement = "FOREDRAGSHOLDER_JZ"|"FOREDRAGSHOLDER_JAVABIN"|"STYRE";

export function allAchievements(): Achievement[] {
    return ["FOREDRAGSHOLDER_JZ", "FOREDRAGSHOLDER_JAVABIN", "STYRE"];
}

export function achievementName(achievement: Achievement): string {
    switch (achievement) {
    case "FOREDRAGSHOLDER_JAVABIN":         return "Foredragsholder JavaBin";
    case "FOREDRAGSHOLDER_JZ":              return "JavaZone foredragsholder";
    case "STYRE":                           return "Styre";
    }
}

export interface CreateHeroData {
    people: Person[];
}

interface ConferenceSpeakerAchievement extends HeroAchievement {
    type: "FOREDRAGSHOLDER_JZ";
    year: number;
    title: string;
}

interface UsergroupSpeakerAchievement extends HeroAchievement {
    type: "FOREDRAGSHOLDER_JAVABIN";
    date: Date;
    title: string;
}

export type BoardMemberRole = "BOARD_MEMBER" | "VICE_CHAIR" | "CHAIR";

export function boardMemberRoleName(role: BoardMemberRole): string {
    switch (role) {
        case "BOARD_MEMBER": return "Board member";
        case "CHAIR": return "Chair";
        case "VICE_CHAIR": return "Vice chair";
    }
}

interface BoardMemberAchivement extends HeroAchievement {
    type: "STYRE";
    year: number;
    role: BoardMemberRole;
}

export type HeroAchievementDetail = ConferenceSpeakerAchievement | UsergroupSpeakerAchievement | BoardMemberAchivement;

interface Heroism {
    achievement: string;
}

interface Consent {
    id: number;
    text: string;
}

export interface HeroProfile {
    profile: Person;
    heroism?: Heroism;
    published?: boolean;
    consent?: Consent;
}

export interface HeroService {
    deleteAchievement(heroId: string, achievementId: string): Promise<void>;
    updateAchievement(heroId: string, achievementId: string, achievement: HeroAchievementDetail): Promise<void>;
    addAchievement(heroId: string, achievement: HeroAchievementDetail): Promise<void>;
    updateHero(heroId: string, update: Partial<Hero>): Promise<void>;
    fetchCreateHeroData(): Promise<CreateHeroData>;
    fetchUserinfo(): Promise<Userinfo>;
    fetchMe(): Promise<HeroProfile>;
    fetchHeroDetails(heroId: string): Promise<Hero>;
    fetchHeroes(): Promise<Hero[]>;
    addHero(hero: Hero): Promise<void>;
    consentToPublish(consentId: number): Promise<void>;
}
