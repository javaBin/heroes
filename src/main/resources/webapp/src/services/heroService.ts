
export interface Hero {
    name: string;
    email: string;
    contribution?: string;
    published: boolean;
}

export interface HeroService {
    fetchMe(): Promise<Hero>;
    fetchHeroes(): Promise<Hero[]>;
    addHero(hero: Hero): Promise<void>;
    consentToPublish(): Promise<void>;
}

export class HeroServiceHttp implements HeroService {
    consentToPublish(): Promise<void> {
        throw new Error("Method not implemented.");
    }
    fetchMe(): Promise<Hero> {
        throw new Error("Method not implemented.");
    }
    addHero(hero: Hero): Promise<void> {
        throw new Error("Method not implemented.");
    }
    async fetchHeroes(): Promise<Hero[]> {
        return [];
    }

}
