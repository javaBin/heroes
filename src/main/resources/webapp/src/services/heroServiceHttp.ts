import { CreateHeroData, Hero, HeroAchievement, HeroProfile, HeroService, Userinfo } from "./api";

export class HeroServiceHttp implements HeroService {
  async deleteAchievement(heroId: string, achievementId: string) {
    await fetch("/api/admin/heroes/" + heroId + "/achievements/" + achievementId, { method: "DELETE" });
  }
  async updateAchievement(heroId: string, achievementId: string, achievement: any) {
    await this.sendJSON("/api/admin/heroes/" + heroId + "/achievements/" + achievementId, achievement, "PUT");
  }
  async addAchievement(heroId: string, achievement: HeroAchievement): Promise<void> {
    await this.sendJSON("/api/admin/heroes/" + heroId + "/achievements", achievement);
  }
  async fetchCreateHeroData(): Promise<CreateHeroData> {
    const response = await fetch("/api/admin/heroes/create");
    if (response.status === 403 || response.status === 401) {
      window.location.href = "/api/login?admin=true";
    }
    return await response.json();
  }

  async fetchHeroDetails(id: string) {
    const response = await fetch("/api/heroes/" + id);
    return await response.json();
  }
  async fetchUserinfo(): Promise<Userinfo> {
    const response = await fetch("/api/userinfo");
    return await response.json();
  }
  async consentToPublish(consentId: number) {
    await this.sendJSON("/api/profiles/mine/consent/1", { consentId });
  }
  async fetchMe(): Promise<HeroProfile> {
    const response = await fetch("/api/profiles/mine");
    if (response.status === 401) {
      window.location.href = "/api/login";
    }
    return await response.json();
  }

  async updateHero(heroId: string, hero: Partial<Hero>) {
    await this.sendJSON("/api/admin/heroes/" + heroId, hero, "PUT");
  }

  async addHero(hero: Hero) {
    await this.sendJSON("/api/admin/heroes", hero);
  }
  async fetchHeroes(): Promise<Hero[]> {
    const response = await fetch("/api/admin/heroes");
    return await response.json();
  }

  async fetchPublicHeroes(): Promise<Hero[]> {
    const response = await fetch("/api/heroes");
    return await response.json();
  }

  async sendJSON(path: string, json: any, method: string = "POST") {
    await fetch(path, {
      body: JSON.stringify(json),
      headers: {
        "Content-type": "application/json"
      },
      method
    });
  }
}
