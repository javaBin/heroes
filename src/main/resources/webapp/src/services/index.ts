import { HeroProfile, HeroService } from "./api";

export { HeroProfile, HeroService };
