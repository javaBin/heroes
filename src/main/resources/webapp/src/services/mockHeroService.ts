import { CreateHeroData, Hero, HeroService } from "./heroService";

interface User {
    admin: boolean;
    email: string;
}

export class MockHeroService implements HeroService {
    currentUser?: User = {
        admin: true,
        email: "johannes@brodwall.com",
    };

    private heroes: Hero[] = [];

    async fetchCreateHeroData() {
        return {
            achievements: [],
            people: [],
        };
    }
    async consentToPublish() {
        this.heroes.find(h => h.email === this.currentUser!.email)!.published = true;
    }

    async fetchUserinfo() {
        return {
            authenticated: !!this.currentUser,
            username: this.currentUser && this.currentUser.email,
        };
    }

    async fetchMe() {
        return this.heroes.find(h => h.email === this.currentUser!.email)!;
    }
    async addHero(hero: Hero) {
        this.heroes.push(hero);
    }
    async fetchHeroes() {
        return this.heroes.filter(h => h.published);
    }

}
