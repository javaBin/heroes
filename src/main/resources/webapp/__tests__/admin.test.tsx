import React from "react";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import { Hero, HeroAchievement, HeroService, Person } from "../src/services/api";
import { MockHeroService } from "../src/services/mockHeroService";

import {
    AddHeroAchievement, AddHeroView, BoardMemberAchievementDetails,
    HeroAchievementList,
    HeroControlPanel, HeroListView,
    HeroView,
    JavaBinSpeakerAchievementDetails,
    JavaZoneSpeakerAchievementDetails,
} from "../src/app/admin/HeroControlPanel";

const fakeHeroService: HeroService = new MockHeroService();

jest.mock("../src/services/heroServiceHttp", () => ({
    HeroServiceHttp: () => fakeHeroService,
}));

// tslint:disable-next-line:no-eval
eval(`global["window"] = {
    location: {
        hash: ""
    },
    addEventListener() {},
    removeEventListener() {}
}`);

function promiseCompletion() {
    return new Promise((resolve) => setImmediate(resolve));
}

describe("HeroControlPanel", () => {

    let heroes: Hero[] = [];

    const people: Person[] = [
        { email: "johannes@example.com", name: "Johannes Brodwall" },
        { email: "alice@example.com", name: "Alice" },
        { email: "bob@example.net", name: "Bob" },
    ];

    beforeEach(() => {
        heroes = [
            {
                achievements: [
                    { id: "13", label: "ABC", type: "FOREDRAGSHOLDER_JAVABIN" },
                    { id: "13", label: "XYZ", type: "FOREDRAGSHOLDER_JAVABIN" },
                ],
                email: "johannes@example.com", id: "1", name: "Johannes Test", published: false,
            },
            {
                achievements: [
                    {id: "11", label: "A great guy!", type: "FOREDRAGSHOLDER_JAVABIN" },
                    {id: "12", label: "A super guy!", type: "STYRE" },
                ],
                email: "test@example.com", id: "2", name: "Some test", published: true,
            },
        ];
        fakeHeroService.addHero = async (hero) => {
            hero.id = "" +  (heroes.length + 1);
            heroes.push(hero);
        };
        fakeHeroService.fetchCreateHeroData = () => Promise.resolve( {people} );
        fakeHeroService.fetchHeroes = () => Promise.resolve(heroes);
        fakeHeroService.updateHero = async (heroId, update) => {
            const index = heroes.findIndex(h => h.id === heroId)!;
            heroes[index] = {...heroes[index], ...update};
        };
        fakeHeroService.addAchievement = (heroId: string, achievement: HeroAchievement) => {
            heroes.find(h => h.id === heroId)!.achievements.push(achievement);
            return Promise.resolve();
        };
        fakeHeroService.updateAchievement = async (heroId, achievementId, update) => {
            const hero = heroes.find(h => h.id === heroId)!;
            const achievementIndex = hero.achievements.findIndex(a => a.id === achievementId);
            hero.achievements[achievementIndex] = {...hero.achievements[achievementIndex], ...update};
        };
        fakeHeroService.deleteAchievement = async (heroId, achievementId) => {
            const hero = heroes.find(h => h.id === heroId)!;
            const achievementIndex = hero.achievements.findIndex(a => a.id === achievementId);
            hero.achievements.splice(achievementIndex);
        };
    });

    it("fakes hero service", async () => {
        const {people} = await fakeHeroService.fetchCreateHeroData();
        expect(people.length).toBe(3);
    });

    it("shows list of current heroes", async () => {
        // tslint:disable-next-line:max-line-length
        const app = renderer.create(<HeroControlPanel heroService={fakeHeroService} prefix="#test" />);
        await promiseCompletion();
        expect(app.toJSON()).toMatchSnapshot();
        expect(app.root.findByType(HeroListView).findByType("ul").findAllByType("a").map(li => li.children[0]))
            .toEqual(["Johannes Test", "Some test"]);
    });

    it("shows hero status", async () => {
        window.location.hash = "#test/heroes/" + heroes[1].id;

        const app = renderer.create(<HeroControlPanel heroService={fakeHeroService} prefix="#test" />);
        await promiseCompletion();
        expect(app.root.findByType(HeroView).findByType("h2").children[0]).toBe(heroes[1].name);
        expect(app.toJSON()).toMatchSnapshot();
    });

    it("shows new hero screen", async () => {
        window.location.hash = "#test/heroes/add";

        const app = renderer.create(<HeroControlPanel heroService={fakeHeroService} prefix="#test" />);
        await promiseCompletion();
        expect(app.root.findByType(AddHeroView).findByType("h2").children[0]).toBe("Add a hero");
        expect(app.toJSON()).toMatchSnapshot();
    });

    it("adds hero from slack list", async () => {
        window.location.hash = "#test/heroes/add";

        const app = renderer.create(<HeroControlPanel heroService={fakeHeroService} prefix="#test" />);
        await promiseCompletion();

        app.root.findByType("select").props.onChange({target: {value: people[1].email}});
        app.root.findByType(AddHeroView).findByType("form").props.onSubmit({preventDefault: jest.fn()});
        await promiseCompletion();

        expect(app.root.instance.state.heroes.map((h: Hero) => h.name)).toContain(people[1].name);
    });

    describe("adds achievement to hero", async () => {
        let app: ReactTestRenderer;
        beforeEach(async () => {
            window.location.hash = "#test/heroes/" + heroes[1].id + "/addAchievement";

            app = renderer.create(<HeroControlPanel heroService={fakeHeroService} prefix="#test" />);
            await promiseCompletion();
        });

        describe("of JavaZone achievement type", () => {
            beforeEach(async () => {
                expect(app.root.findAll(c => true).map(c => c.type)).toContain(AddHeroAchievement);
                app.root.findByType(AddHeroAchievement).instance
                .setState({achievementType: "FOREDRAGSHOLDER_JZ"});
                await promiseCompletion();
            });

            it("shows achievement form", async () => {
                app.root.findByType(JavaZoneSpeakerAchievementDetails);
                expect(app.toJSON()).toMatchSnapshot();
            });

            it("creates new achievement", async () => {
                const form = app.root.findByType(AddHeroAchievement).findByType("form");
                const titleInput = form.findByType("input");
                const submitButton = form.findByType("button");
                titleInput.props.onChange({ target: { value: "My Talk"}});
                submitButton.props.onClick({preventDefault: jest.fn()});
                await promiseCompletion();

                expect(app.root.instance.state.heroes[1].achievements
                    .map((a: any) => a.title)).toContain("My Talk");
            });
        });

        it("shows JavaBin achievement type", async () => {
            app.root.findByType(AddHeroAchievement).instance
                .setState({achievementType: "FOREDRAGSHOLDER_JAVABIN"});
            await promiseCompletion();
            app.root.findByType(JavaBinSpeakerAchievementDetails);
            expect(app.toJSON()).toMatchSnapshot();
        });

        it("shows board member achievement type", async () => {
            app.root.findByType(AddHeroAchievement).instance
                .setState({achievementType: "STYRE"});
            await promiseCompletion();
            app.root.findByType(BoardMemberAchievementDetails);
            expect(app.toJSON()).toMatchSnapshot();
        });
    });

    it("shows hero edit view", async () => {
        window.location.hash = "#test/heroes/" + heroes[1].id + "/edit";

        const app = renderer.create(<HeroControlPanel heroService={fakeHeroService} prefix="#test" />);
        await promiseCompletion();
        const [displayName, email, twitter] = app.root.findByType("form").findAllByType("input");
        expect(displayName.props.value).toEqual(heroes[1].name);
        expect(email.props.value).toEqual(heroes[1].email);
        expect(twitter.props.value).toEqual(heroes[1].twitter);

        expect(app.toJSON()).toMatchSnapshot();
    });

    it("udpates hero name", async () => {
        window.location.hash = "#test/heroes/" + heroes[0].id + "/edit";

        const app = renderer.create(<HeroControlPanel heroService={fakeHeroService} prefix="#test" />);
        await promiseCompletion();
        const [displayName] = app.root.findByType("form").findAllByType("input");
        await displayName.props.onChange({target: {value: "Updated name"}});
        app.root.findByType("form").props.onSubmit({preventDefault: jest.fn()});

        await promiseCompletion();
        expect(heroes[0].name).toEqual("Updated name");

        expect(app.toJSON()).toMatchSnapshot();
    });

    it("lists achievements", async () => {
        window.location.hash = "#test/heroes/" + heroes[1].id;

        const app = renderer.create(<HeroControlPanel heroService={fakeHeroService} prefix="#test" />);

        await promiseCompletion();
        expect(app.root.findByType(HeroAchievementList).findAllByType("li").map(li => li.children[0]))
            .toEqual(heroes[1].achievements.map(a => a.label));
        expect(app.toJSON()).toMatchSnapshot();
    });

    it("deletes achievement", async () => {
        window.location.hash = "#test/heroes/" + heroes[0].id;
        const deleted = heroes[0].achievements[0].label;

        const app = renderer.create(<HeroControlPanel heroService={fakeHeroService} prefix ="#test" />);
        await promiseCompletion();
        app.root.findByType(HeroAchievementList).findAllByProps({className: "deleteAchievementLink"})[0]
            .props.onClick({preventDefault: jest.fn});
        await promiseCompletion();
        expect(heroes[0].achievements.map(a => a.label)).not.toContain(deleted);
        expect(app.toJSON()).toMatchSnapshot();
    });

    it("shows achievement update", async () => {
        window.location.hash = "#test/heroes/" + heroes[1].id + "/achievement/" + heroes[1].achievements[1].id;

        const app = renderer.create(<HeroControlPanel heroService={fakeHeroService} prefix="#test" />);
        await promiseCompletion();
        app.root.findByType(BoardMemberAchievementDetails);
        expect(app.toJSON()).toMatchSnapshot();
    });

    it("updates speaker achievement", async () => {
        window.location.hash = "#test/heroes/" + heroes[1].id + "/achievement/" + heroes[1].achievements[0].id;

        const app = renderer.create(<HeroControlPanel heroService={fakeHeroService} prefix="#test" />);
        await promiseCompletion();
        app.root.findByType(JavaBinSpeakerAchievementDetails);
        const achievementView = app.root.findByType(JavaBinSpeakerAchievementDetails);
        const [titleInput, dateInput] = achievementView.findAllByType("input");
        await titleInput.props.onChange({target: {value: "Updated title"}});
        await dateInput.props.onChange({target: {value: "2018/10/15"}});
        achievementView.findByType("button").props.onClick({preventDefault: jest.fn()});

        expect(heroes[1].achievements[0]).toHaveProperty("title", "Updated title");
        expect(app.toJSON()).toMatchSnapshot();
    });
});
