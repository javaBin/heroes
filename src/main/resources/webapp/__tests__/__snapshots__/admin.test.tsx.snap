// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroControlPanel adds achievement to hero of JavaZone achievement type shows achievement form 1`] = `
Array [
  <h3>
    <a
      href="#test"
    >
      Back
    </a>
  </h3>,
  <h2>
    Some test
  </h2>,
  <h3>
    Please Add achievement
  </h3>,
  <form>
    <label>
      Achievement:
      <select
        autoFocus={true}
        onChange={[Function]}
      >
        <option />
        <option
          value="FOREDRAGSHOLDER_JZ"
        >
          JavaZone foredragsholder
        </option>
        <option
          value="FOREDRAGSHOLDER_JAVABIN"
        >
          Foredragsholder JavaBin
        </option>
        <option
          value="STYRE"
        >
          Styre
        </option>
      </select>
    </label>
    <div>
      <h4>
        JavaZone foredragsholder
      </h4>
      <label>
        JavaZone year
        <select
          autoFocus={true}
          onChange={[Function]}
          value="2018"
        >
          <option
            value={2018}
          >
            2018
          </option>
          <option
            value={2017}
          >
            2017
          </option>
          <option
            value={2016}
          >
            2016
          </option>
          <option
            value={2015}
          >
            2015
          </option>
          <option
            value={2014}
          >
            2014
          </option>
          <option
            value={2013}
          >
            2013
          </option>
          <option
            value={2012}
          >
            2012
          </option>
          <option
            value={2011}
          >
            2011
          </option>
          <option
            value={2010}
          >
            2010
          </option>
          <option
            value={2009}
          >
            2009
          </option>
          <option
            value={2008}
          >
            2008
          </option>
        </select>
      </label>
      <label>
        Title:
        <input
          onChange={[Function]}
          value=""
        />
      </label>
      <button
        disabled={true}
        onClick={[Function]}
      >
        Submit
      </button>
    </div>
    <a
      href="#test/heroes/2"
    >
      Back
    </a>
  </form>,
]
`;

exports[`HeroControlPanel adds achievement to hero shows JavaBin achievement type 1`] = `
Array [
  <h3>
    <a
      href="#test"
    >
      Back
    </a>
  </h3>,
  <h2>
    Some test
  </h2>,
  <h3>
    Please Add achievement
  </h3>,
  <form>
    <label>
      Achievement:
      <select
        autoFocus={true}
        onChange={[Function]}
      >
        <option />
        <option
          value="FOREDRAGSHOLDER_JZ"
        >
          JavaZone foredragsholder
        </option>
        <option
          value="FOREDRAGSHOLDER_JAVABIN"
        >
          Foredragsholder JavaBin
        </option>
        <option
          value="STYRE"
        >
          Styre
        </option>
      </select>
    </label>
    <h3>
      JavaBin usergroup speaker
    </h3>
    <label>
      Title:
      <input
        onChange={[Function]}
        value=""
      />
    </label>
    <label>
      Date:
      <input
        onChange={[Function]}
        type="date"
        value=""
      />
    </label>
    <button
      disabled={true}
      onClick={[Function]}
    >
      Submit
    </button>
    <a
      href="#test/heroes/2"
    >
      Back
    </a>
  </form>,
]
`;

exports[`HeroControlPanel adds achievement to hero shows board member achievement type 1`] = `
Array [
  <h3>
    <a
      href="#test"
    >
      Back
    </a>
  </h3>,
  <h2>
    Some test
  </h2>,
  <h3>
    Please Add achievement
  </h3>,
  <form>
    <label>
      Achievement:
      <select
        autoFocus={true}
        onChange={[Function]}
      >
        <option />
        <option
          value="FOREDRAGSHOLDER_JZ"
        >
          JavaZone foredragsholder
        </option>
        <option
          value="FOREDRAGSHOLDER_JAVABIN"
        >
          Foredragsholder JavaBin
        </option>
        <option
          value="STYRE"
        >
          Styre
        </option>
      </select>
    </label>
    <h3>
      JavaBin board member
    </h3>
    <label>
      Role
      <select
        onChange={[Function]}
        value="BOARD_MEMBER"
      >
        <option
          value="BOARD_MEMBER"
        >
          Board member
        </option>
        <option
          value="VICE_CHAIR"
        >
          Vice chair
        </option>
        <option
          value="CHAIR"
        >
          Chair
        </option>
      </select>
    </label>
    <label>
      Elected year
      <select
        onChange={[Function]}
        value="2018"
      >
        <option
          value="2018"
        >
          2018
        </option>
        <option
          value="2017"
        >
          2017
        </option>
        <option
          value="2016"
        >
          2016
        </option>
        <option
          value="2015"
        >
          2015
        </option>
        <option
          value="2014"
        >
          2014
        </option>
        <option
          value="2013"
        >
          2013
        </option>
        <option
          value="2012"
        >
          2012
        </option>
        <option
          value="2011"
        >
          2011
        </option>
        <option
          value="2010"
        >
          2010
        </option>
        <option
          value="2009"
        >
          2009
        </option>
        <option
          value="2008"
        >
          2008
        </option>
      </select>
    </label>
    <button
      disabled={false}
      onClick={[Function]}
    >
      Submit
    </button>
    <a
      href="#test/heroes/2"
    >
      Back
    </a>
  </form>,
]
`;

exports[`HeroControlPanel deletes achievement 1`] = `
Array [
  <h3>
    <a
      href="#test"
    >
      Back
    </a>
  </h3>,
  <h2>
    Johannes Test
  </h2>,
  <ul>
    <li>
      Email: 
      johannes@example.com
    </li>
    <li>
      Twitter: 
    </li>
    <li>
      <a
        href="#test/heroes/1/edit"
      >
        Update
      </a>
    </li>
  </ul>,
  <h3>
    Achievements
  </h3>,
  <ul />,
  <p>
    <a
      href="#test/heroes/1/addAchievement"
    >
      Add achievement
    </a>
  </p>,
]
`;

exports[`HeroControlPanel lists achievements 1`] = `
Array [
  <h3>
    <a
      href="#test"
    >
      Back
    </a>
  </h3>,
  <h2>
    Some test
  </h2>,
  <ul>
    <li>
      Email: 
      test@example.com
    </li>
    <li>
      Twitter: 
    </li>
    <li>
      <a
        href="#test/heroes/2/edit"
      >
        Update
      </a>
    </li>
  </ul>,
  <h3>
    Achievements
  </h3>,
  <ul>
    <li>
      A great guy!
       [
      <a
        href="#test/heroes/2/achievement/11"
      >
        Edit
      </a>
      ] [
      <a
        className="deleteAchievementLink"
        href="#"
        onClick={[Function]}
      >
        Delete
      </a>
      ]
    </li>
    <li>
      A super guy!
       [
      <a
        href="#test/heroes/2/achievement/12"
      >
        Edit
      </a>
      ] [
      <a
        className="deleteAchievementLink"
        href="#"
        onClick={[Function]}
      >
        Delete
      </a>
      ]
    </li>
  </ul>,
  <p>
    <a
      href="#test/heroes/2/addAchievement"
    >
      Add achievement
    </a>
  </p>,
]
`;

exports[`HeroControlPanel shows achievement update 1`] = `
Array [
  <h3>
    <a
      href="#test"
    >
      Back
    </a>
  </h3>,
  <h2>
    Some test
  </h2>,
  <form>
    <h3>
      JavaBin board member
    </h3>
    <label>
      Role
      <select
        onChange={[Function]}
        value="BOARD_MEMBER"
      >
        <option
          value="BOARD_MEMBER"
        >
          Board member
        </option>
        <option
          value="VICE_CHAIR"
        >
          Vice chair
        </option>
        <option
          value="CHAIR"
        >
          Chair
        </option>
      </select>
    </label>
    <label>
      Elected year
      <select
        onChange={[Function]}
        value="2018"
      >
        <option
          value="2018"
        >
          2018
        </option>
        <option
          value="2017"
        >
          2017
        </option>
        <option
          value="2016"
        >
          2016
        </option>
        <option
          value="2015"
        >
          2015
        </option>
        <option
          value="2014"
        >
          2014
        </option>
        <option
          value="2013"
        >
          2013
        </option>
        <option
          value="2012"
        >
          2012
        </option>
        <option
          value="2011"
        >
          2011
        </option>
        <option
          value="2010"
        >
          2010
        </option>
        <option
          value="2009"
        >
          2009
        </option>
        <option
          value="2008"
        >
          2008
        </option>
      </select>
    </label>
    <button
      disabled={false}
      onClick={[Function]}
    >
      Submit
    </button>
  </form>,
]
`;

exports[`HeroControlPanel shows hero edit view 1`] = `
Array [
  <h3>
    <a
      href="#test"
    >
      Back
    </a>
  </h3>,
  <h2>
    Some test
  </h2>,
  <form
    onSubmit={[Function]}
  >
    <h3>
      Information
    </h3>
    <ul>
      <li>
        Display name:
        <input
          autoFocus={true}
          onChange={[Function]}
          value="Some test"
        />
      </li>
      <li>
        Email address:
        <input
          onChange={[Function]}
          value="test@example.com"
        />
      </li>
      <li>
        Twitter handle:
        <input
          onChange={[Function]}
        />
      </li>
    </ul>
    <button>
      Lagre
    </button>
    <a
      href="#test/heroes/2"
    >
      Back
    </a>
  </form>,
]
`;

exports[`HeroControlPanel shows hero status 1`] = `
Array [
  <h3>
    <a
      href="#test"
    >
      Back
    </a>
  </h3>,
  <h2>
    Some test
  </h2>,
  <ul>
    <li>
      Email: 
      test@example.com
    </li>
    <li>
      Twitter: 
    </li>
    <li>
      <a
        href="#test/heroes/2/edit"
      >
        Update
      </a>
    </li>
  </ul>,
  <h3>
    Achievements
  </h3>,
  <ul>
    <li>
      A great guy!
       [
      <a
        href="#test/heroes/2/achievement/11"
      >
        Edit
      </a>
      ] [
      <a
        className="deleteAchievementLink"
        href="#"
        onClick={[Function]}
      >
        Delete
      </a>
      ]
    </li>
    <li>
      A super guy!
       [
      <a
        href="#test/heroes/2/achievement/12"
      >
        Edit
      </a>
      ] [
      <a
        className="deleteAchievementLink"
        href="#"
        onClick={[Function]}
      >
        Delete
      </a>
      ]
    </li>
  </ul>,
  <p>
    <a
      href="#test/heroes/2/addAchievement"
    >
      Add achievement
    </a>
  </p>,
]
`;

exports[`HeroControlPanel shows list of current heroes 1`] = `
Array [
  <h2>
    Here are all the heroes
  </h2>,
  <ul>
    <li>
      <a
        href="#test/heroes/1"
      >
        Johannes Test
      </a>
    </li>
    <li>
      <a
        href="#test/heroes/2"
      >
        Some test
      </a>
    </li>
  </ul>,
  <a
    href="#test/heroes/add"
  >
    Add a hero
  </a>,
]
`;

exports[`HeroControlPanel shows new hero screen 1`] = `
<form
  onSubmit={[Function]}
>
  <h2>
    Add a hero
  </h2>
  <div>
    <label>
      Select from slack:
      <select
        autoFocus={true}
        onChange={[Function]}
      >
        <option />
        <option
          value="johannes@example.com"
        >
          Johannes Brodwall
           &lt;
          johannes@example.com
          &gt;
        </option>
        <option
          value="alice@example.com"
        >
          Alice
           &lt;
          alice@example.com
          &gt;
        </option>
        <option
          value="bob@example.net"
        >
          Bob
           &lt;
          bob@example.net
          &gt;
        </option>
      </select>
    </label>
  </div>
  <div>
    <label>
      Name:
      <input
        onChange={[Function]}
      />
    </label>
  </div>
  <div>
    <label>
      Email:
      <input
        onChange={[Function]}
      />
    </label>
  </div>
  <div>
    <label>
      Twitter:
      <input
        onChange={[Function]}
      />
    </label>
  </div>
  <button>
    Submit
  </button>
  <button
    onClick={[Function]}
  >
    Cancel
  </button>
</form>
`;

exports[`HeroControlPanel udpates hero name 1`] = `
Array [
  <h3>
    <a
      href="#test"
    >
      Back
    </a>
  </h3>,
  <h2>
    Updated name
  </h2>,
  <form
    onSubmit={[Function]}
  >
    <h3>
      Information
    </h3>
    <ul>
      <li>
        Display name:
        <input
          autoFocus={true}
          onChange={[Function]}
          value="Updated name"
        />
      </li>
      <li>
        Email address:
        <input
          onChange={[Function]}
          value="johannes@example.com"
        />
      </li>
      <li>
        Twitter handle:
        <input
          onChange={[Function]}
        />
      </li>
    </ul>
    <button>
      Lagre
    </button>
    <a
      href="#test/heroes/1"
    >
      Back
    </a>
  </form>,
]
`;

exports[`HeroControlPanel updates speaker achievement 1`] = `
Array [
  <h3>
    <a
      href="#test"
    >
      Back
    </a>
  </h3>,
  <h2>
    Some test
  </h2>,
  <form>
    <h3>
      JavaBin usergroup speaker
    </h3>
    <label>
      Title:
      <input
        onChange={[Function]}
        value="Updated title"
      />
    </label>
    <label>
      Date:
      <input
        onChange={[Function]}
        type="date"
        value="2018/10/15"
      />
    </label>
    <button
      disabled={false}
      onClick={[Function]}
    >
      Submit
    </button>
  </form>,
]
`;
